### ISSUE TITLE
Discussion: scalar expression with array return type for each SRF?

### ISSUE THREAD

**Issue by st1page** [risingwavelabs/risingwave#11745]
in PG, SRF(Set Returning Function) is used to return more than one row. Each input row will generate multiple rows in the result set, so, the result set is grouped by the original input row naturally. For example:
```
dev=# create table t (s text, id int primary key);
CREATE TABLE
dev=# insert into t values ('a a a1 a2', 1), ('b b1 b2 a', 2);
INSERT 0 2
dev=# select id, regexp_matches(s, 'a', 'g') from t;
 id | regexp_matches 
----+----------------
  1 | {a}
  1 | {a}
  1 | {a}
  1 | {a}
  2 | {a}
(5 rows)
dev=# select count(*), id from (
    select id, regexp_matches(s, 'a', 'g') from t
) m group by id;
 count | id 
-------+----
     1 |  2
     4 |  1
(2 rows)
```
It is because in the classical relational model, high-dimensional data can only be stored in one relation. But RW support semi sturcture with array and struct. So can we give all SRF an array-return scalar version function? Also, we can reduce many work to reuse the implementation of SRF.
e.g.
```SQL
dev=# select id, regexp_matches_arr(s, 'a', 'g') from t;
regexp_matches_arr | id 
-------------------+---------
 {"a","a","a","a"} |       1
(1 row)
```

**Comment by fuyufjh** [comment]
LGTM. By the way, I found a similar requirements [here](https://stackoverflow.com/questions/16609554/retrieve-results-from-regexp-matches-as-one-row)

Another question is: shall we provide both (SRF & scalar), or provide scalar version only? If we can "reuse the implementation of SRF", I prefer to provide both.

**Comment by xiangjinwu** [comment]
As the link mentioned above, it is possible to use `array_agg` or `array(subquery)` in PostgreSQL:
```
test=# select id, (select array_agg(r) from regexp_matches(s, 'a', 'g') r) from t;
 id |     array_agg     
----+-------------------
  1 | {{a},{a},{a},{a}}
  2 | {{a}}
(2 rows)

test=# select id, array(select regexp_matches(s, 'a', 'g')) from t;
 id |       array       
----+-------------------
  1 | {{a},{a},{a},{a}}
  2 | {{a}}
(2 rows)
```

Also note the result is 2d array because regexp can contain multiple groups.

Basically, `unnest` and `array_agg` can be used to convert between setof and array. That being said, PostgreSQL also provides some functions in 2 forms:
* `string_to_array` and `string_to_table`
* `regexp_split_to_array` and `regexp_split_to_table`

**Comment by st1page** [comment]
> As the link mentioned above, it is possible to use array_agg or array(subquery) in PostgreSQL:
`array(subquery)` LGTM, but it might be more flexible and harder than just supporting `array(SRF)` :thinking:

**Comment by chenzl25** [comment]
It seems `array(subquery)` can always be replaced with `array_agg`. 
For example: 
use `select (select array_agg(x) from t2 where y = t.v1) from t;` 
to replace `select array(select x from t2 where y = t.v1) from t;`.

**Comment by st1page** [comment]
> It seems `array(subquery)` can always be replaced with `array_agg`. For example: use `select (select array_agg(x) from t2 where y = t.v1) from t;` to replace `select array(select x from t2 where y = t.v1) from t;`.

Yes, but prefer to implement it in execution to avoid unnecessary join and group agg.

**Comment by fuyufjh** [comment]
> `array(subquery)` LGTM, but it might be more flexible and harder than just supporting `array(SRF)` 🤔 because we can not execute nested subquery in streaming execution. c.c. @chenzl25

I'm wondering whether it's **possible** to implement `array(...)`? Particularly, can `array(...)` be evaluated not by nested-loop?

**Comment by fuyufjh** [comment]
A random idea:

An ad-hoc approach is to introduce a special path for `ARRAY(...)`: if the inner only contains stateless operators like Project, ProjectSet and Filter, replace the `ARRAY(...)` with a stateless `Apply` operator, which executes a simple batch query (no TableScanm, no shuffle) every time to get the results.

**Comment by chenzl25** [comment]
> A random idea:
An ad-hoc approach is to introduce a special path for ARRAY(...): if the inner only contains stateless operators like Project, ProjectSet and Filter, replace the ARRAY(...) with a stateless Apply operator, which executes a simple batch query (no TableScan, no shuffle) every time to get the results.


Let's take a look at the logical plan generated by our planner of the following SQL.
```sql
explain (trace) select id, (select array_agg(x) from regexp_matches(s, 'a', 'g') x) from t;
```


If the RHS of the `Apply` only depends on the LHS, theoretically, it can execute a simple batch query to calculate the result.

```
 LogicalProject { exprs: [t.id, array_agg(regexp_matches)] }
 └─LogicalApply { type: LeftOuter, on: true, correlated_id: 1, max_one_row: true }
   ├─LogicalScan { table: t, columns: [s, id] }
   └─LogicalProject { exprs: [array_agg(regexp_matches)] }
     └─LogicalAgg { aggs: [array_agg(regexp_matches)] }
       └─LogicalProject { exprs: [regexp_matches] }
         └─LogicalTableFunction { table_function: RegexpMatches(CorrelatedInputRef { index: 0, correlated_id: 1 }, 'a':Varchar, 'g':Varchar) }
```

**Comment by fuyufjh** [comment]
When I just started to write an RFC about the idea on the above comments, I realized that SRF (Set Returning Function) seems to be a PG-specific thing and most users may not be familiar with it. Now I tend to keep it simple and stupid and just introduce an array version for SRFs instead, such as `regexp_matches_array()`, which is quite clear and understandable to me.
